Slot game.part5

Fedrogan opened the pull request:
5th task completed.
Bonus game (freespins) and PopUps added.

Commit:
5th task temp

Commit:
Bonus game (free spins) added. (temp.)
TODO:
1. FreeSpins counter
2. PopUp

Commit:
temp

Commit:
SLOT GAME. Part 5
Bonus game and popups added.

MPRidotto commented on Assets/Scripts/BonusGameController.cs:
```diff
@@ -0,0 +1,191 @@
+    }
+    private void Start()
+    {
+        popUpsContainer.BonusGameStartPopUpShown += StartBonusGame;
```
Не нравится привязка начала фриспинов к popUpsContainer и показу попапа. Ведь показ попапа не является истинной причиной, а является всего лишь следствием.

Fedrogan replied:
GameController принимает от winLinesChecker количество найденных на финальном экране скаттеров и через gameModel.UpdateGame обновляет стэйт.

MPRidotto commented on Assets/Scripts/BonusGameController.cs:
```diff
@@ -0,0 +1,191 @@
+    private void Start()
+    {
+        popUpsContainer.BonusGameStartPopUpShown += StartBonusGame;
+        popUpsContainer.TotalPrizePopUpShown += FinishBonusGame;
```
Тоже, что и с началом фриспинов.

Fedrogan replied:
То же, что и с началом фриспинов =))

MPRidotto commented on Assets/Scripts/BonusGameController.cs:
```diff
@@ -0,0 +1,191 @@
+    }
+
+
+    private void OnAllReelsStopped()
```
Не очень удачное место для метода в этом классе, т.к. все, что происходит по остановке рилов (анимации и т.д.), должно контролироваться в том месте, где это будет происходить, как во время фс, так и во время обычных вращений.

Fedrogan replied:
BonusGameController удален, теперь в игре один главный контроллер

MPRidotto commented on Assets/Scripts/GameType.cs:
```diff
@@ -0,0 +1,5 @@
+﻿public enum GameType
+{
+    Ordinary,
+    Bonus
```
Лучше все-таки обозвать FreeSpins, т.к. в будущем ты увидишь, что бонусные игры, это немного другое

Fedrogan replied:
Вроде нигде не забыл поменять

MPRidotto commented on Assets/Scripts/PopUpsContainer.cs:
```diff
@@ -0,0 +1,102 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+public class PopUpsContainer : MonoBehaviour
```
А если в игре будет 50 попапов, то они также будут обрабатываться с анимациями и подпиской на их кнопки в этом классе?

MPRidotto commented:
Основные замечания касаются того, что рилы и анимации должны функционировать независимо от того, это фс или нет. Еще не понрмавился класс popUpsContainer, получился он слишком нерасширяемым, с одной стороны, если нет требований, что будет функционал дополнен, то и не надо заморачиваться, но тут другой случай, уж слишком нарушены принципы SOLID.

Commit:
Game type Bonus replaced with FreeSpins

Commit:
beforeMVC

Commit:
SLOT GAME. Part 5. Conversations resolving.
1. start and end of free spins are no longer tied to popups
2. each popup has a script that implements a simple interface IPopup
3. responsibilities are separated and ordered:
Model - GameModel - stores the state and basic elements of the game(GameType, totalWinning, freeSpinsLeft)
Controller - GameController(Main controller) - handles events of other smaller controllers, changes game state, interacts with view objects through their methods

Fedrogan commented:
SLOT GAME. Part 5. Conversations resolving.
1. start and end of free spins are no longer tied to popups
2. each popup has a script that implements a simple interface IPopup
3. responsibilities are separated and ordered:
Model - GameModel - stores the state and basic elements of the game(GameType, totalWinning, freeSpinsLeft)
Controller - GameController(Main controller) - handles events of other smaller controllers, changes game state, interacts with view objects through their methods
